### ISSUE TITLE
theano.function generate a shared library with a C example

### ISSUE THREAD

**Issue by nouiz** [Theano/Theano#2722]
This is a follow up from gh-1408. We should not focus on windows as gh-1408, but we can take the branch started there as it was working for some case on Linux.

The branch that contain the current prototype:
https://github.com/nouiz/Theano/tree/shared_lib_rebased

Problem: It would be very useful to generate a shared library from a Theano function. This would allow to reuse it in other program and on embdede system more easily.

- [ ] Rebase the branch
- [ ] Make a user interface to enable/not enable that feature
- [ ] Bring the prototype to a working version without adding new feature.
- [ ] Document it.
- [ ] Add support for scalar constant value in the graph.
- [ ] Make a configuration option to enable/disable GC of intermediate results.
- [ ] Make an interface to support shared variables.

later:
- [ ] To make it work on Windows, we need to back-port some c code that use C99 features.

**Comment by cancan101** [comment]
Is this related: https://github.com/Theano/Theano/issues/2271 ?

**Comment by nouiz** [comment]
Yes, it mostly tell more information on the follow up from gh-1408. I updated 2271 to reference this one instead.

**Comment by nouiz** [comment]
Some update, a rebase is available in this branch: https://github.com/AlOa/Theano/tree/embed_theano

**Comment by mpslxz** [comment]
@nouiz Any updates about this issue? I was thinking about starting it. Do you think it is still worth doing it?

**Comment by nouiz** [comment]
I still see a need for that, but it isn't an easy problem and we currently don't have time to help/review work on this. This week, someone started to work here, so maybe we could handle review/guide on this, but it isn't sure.

We have the deep learning montreal summer school for the next 2 weeks, so now isn't a good time.

I made a prototype many years ago and @AlOa continued it and rebased it a few times: https://github.com/Theano/Theano/pull/3350, but I didn't got enough time to review/guide him well enough. What is your timeline? What exactly do you want?

**Comment by mpslxz** [comment]
Oh I understand that you guys are very busy because of the summer school. My schedule is packed too but I'd love to make some time for this and I definitely agree that this is not an easy thing to do. I was thinking about having a "converter" that traverses the computation graph and generates the equivalent in say MKL or some parallel library. However, I liked the idea of shared library more.
